src/devices/bus/multibus/robotron_k7070.cpp

```diff
@@ -0,0 +1,408 @@
+
+	// V.24 port for tablet
+	RS232_PORT(config, m_serial[0], default_rs232_devices, "loopback");
+	m_serial[0]->rxd_handler().set(m_sio, FUNC(z80sio_device::rxa_w));
+	m_serial[0]->dcd_handler().set(m_sio, FUNC(z80sio_device::dcda_w));
+	m_serial[0]->cts_handler().set(m_sio, FUNC(z80sio_device::ctsa_w));
+
+	// IFSS (current loop) port for keyboard (unused on A7150)
+	RS232_PORT(config, m_serial[1], default_rs232_devices, "loopback");
+	m_serial[1]->rxd_handler().set(m_sio, FUNC(z80sio_device::rxb_w));
+
+	screen_device &screen(SCREEN(config, "screen", SCREEN_TYPE_RASTER));
+	screen.set_color(rgb_t::green());
+	screen.set_raw(XTAL(16'000'000), 737, 0, 640, 431, 0, 400);
+	screen.set_screen_update(FUNC(robotron_k7070_device::screen_update_k7072));
+	screen.set_palette(m_palette);
+	screen.screen_vblank().set([this] (bool state) { if (m_nmi_enable) m_cpu->set_input_line(INPUT_LINE_NMI, state); });
+
+	PALETTE(config, m_palette, palette_device::MONOCHROME);
+}
+
+void robotron_k7070_device::cpu_mem(address_map &map)
+{
+	map.unmap_value_high();
+
+	map(0x0000, 0x1fff).view(m_view_lo);
+	m_view_lo[0](0x0000, 0x1fff).rom().region("eprom", 0);
+	m_view_lo[1](0x0000, 0x1fff).ram().share("kgs_ram0");
+
+	map(0x2000, 0x7fff).ram().share("kgs_ram1");
+
+// FIXME handle IML=0 and MSEL=1 (no access to VRAM)
+	map(0x8000, 0xffff).view(m_view_hi);
+	m_view_hi[0](0x8000, 0xffff).ram().share("kgs_ram2");
+	m_view_hi[1](0x8000, 0xffff).ram().share("vram0");
+}
+
+void robotron_k7070_device::cpu_pio(address_map &map)
+{
+	map.unmap_value_high();
+	map.global_mask(0xff);
+
+	map(0x0000, 0x0003).rw(m_ctc, FUNC(z80ctc_device::read), FUNC(z80ctc_device::write));
+	map(0x0008, 0x000b).rw(m_sio, FUNC(z80sio_device::ba_cd_r), FUNC(z80sio_device::ba_cd_w));
+	map(0x0010, 0x0017).rw(FUNC(robotron_k7070_device::kgs_host_r), FUNC(robotron_k7070_device::kgs_host_w)); // p. 11 of KGS-K7070.pdf
+
+	// p. 6 of ABG-K7072.pdf
+	map(0x0020, 0x0021).w(FUNC(robotron_k7070_device::abg_addr_w));
+	map(0x0022, 0x0022).w(FUNC(robotron_k7070_device::abg_func_w));
+	map(0x0023, 0x0023).w(FUNC(robotron_k7070_device::abg_split_w));
+	map(0x0030, 0x003f).noprw(); // palette register
+
+	map(0x0080, 0x0080).w(FUNC(robotron_k7070_device::abg_misc_w));
+}
+
+void robotron_k7070_device::palette_init(palette_device &palette)
+{
+	// 0=black, 1=dim, 2=normal, 3=bold
+
+	palette.set_pen_color(0, rgb_t::black());
+	palette.set_pen_color(1, rgb_t(0x80, 0x80, 0x80));
+	palette.set_pen_color(2, rgb_t(0xc0, 0xc0, 0xc0));
+	palette.set_pen_color(3, rgb_t::white());
+}
+
+uint16_t robotron_k7070_device::io_r(offs_t offset, uint16_t mem_mask)
+{
+	uint8_t data = 0;
+
+	switch (offset)
+	{
+	case 0:
+		data = m_kgs_ctrl;
+		break;
+
+	case 1:
+		data = m_kgs_datai;
+		m_kgs_ctrl &= ~KGS_ST_OBF;
+		break;
+	}
+
+	if (offset)
+	LOG("%s: KGS %d == %02x '%c'\n", machine().describe_context(), offset, data,
+			 (data > 0x1f && data < 0x7f) ? data : 0x20);
```

Comment:
Please indent controlled statements by one level.  Also, please check `machine().side_effects_disabled()` when logging in a read handler to avoid spam when a debugger window is open.